src/Chainweb/Miner/Core.hs

```diff
@@ -56,60 +73,57 @@ usePowHash Testnet02 f = f $ Proxy @SHA512t_256
 --
 -- TODO: Remove the `Proxy`?
 --
-mine :: forall a. HashAlgorithm a => Proxy a -> BlockHeader -> IO BlockHeader
-mine _ h = BA.withByteArray initialTargetBytes $ \trgPtr -> do
+mine
+  :: forall a. HashAlgorithm a
+  => Proxy a
+  -> TargetBytes
+  -> NonceBytes
+  -> HeaderBytes
+  -> IO HeaderBytes
+mine _ (TargetBytes tbytes) nonce (HeaderBytes hbytes) = BA.withByteArray tbytes $ \trgPtr -> do
     !ctx <- hashMutableInit @a
-    bytes <- BA.copy initialBytes $ \buf ->
+    fmap HeaderBytes . BA.copy hbytes $ \buf ->
         allocaBytes (powSize :: Int) $ \pow -> do
 
             -- inner mining loop
             --
-            let go !i !n = do
+            let go !n = do
                     -- Compute POW hash for the nonce
                     injectNonce n buf
                     hash ctx buf pow
 
                     -- check whether the nonce meets the target
                     fastCheckTarget trgPtr (castPtr pow) >>= \case
                         True -> pure ()
-                        False -> go (succ i) (succ n)
+                        False -> go (succ n)
 
             -- Start inner mining loop
-            go (0 :: Int) $ _blockNonce h
-
-    -- On success: deserialize and return the new BlockHeader
-    runGet decodeBlockHeaderWithoutHash bytes
+            go nonce
   where
-    initialBytes :: B.ByteString
-    !initialBytes = runPutS $ encodeBlockHeaderWithoutHash h
-
-    initialTargetBytes :: B.ByteString
-    !initialTargetBytes = runPutS $ encodeHashTarget target
+    -- tbytes :: B.ByteString
+    -- !tbytes = runPutS $ encodeHashTarget target
 
     bufSize :: Int
-    !bufSize = B.length initialBytes
-
-    target :: HashTarget
-    !target = _blockTarget h
+    !bufSize = B.length hbytes
 
     powSize :: Int
-    !powSize = int $ hashDigestSize @a undefined
```

Comment:
`-Widentities` detects this as a spurious conversion, but only when supplied as `fromIntegral`. The over-sight goes undetected with just `int`.